### PR TITLE
MGMT-3862 update KubeAPI Image with recent StaticIPConfig changes

### PR DESCRIPTION
**Pull request by nmagnezi**

### CODE REVIEW

**Comment by nmagnezi** [comment]
/hold
testing

**Comment by nmagnezi** [comment]
Tested:
====

**Build**
* `skipper make manifests ENABLE_KUBE_API=true`, followed by `skipper make all` ==> built ok.

**Create resources**
1. `skipper make deploy-test ENABLE_KUBE_API=true` (switch auth off)
2. kubectl -n assisted-installer apply -f [pullsecret.yaml]( https://github.com/nmagnezi/assisted-service-crs/blob/d2eb98f937e93ce35e4507eb84a428b17a11c58b/pullsecret.yaml)
3. kubectl -n assisted-installer apply -f [cluster.yaml](https://github.com/nmagnezi/assisted-service-crs/blob/cc900a4c39295c0b9204ff3bb25f5ed940f402a9/cluster.yaml)
4. kubectl -n assisted-installer apply -f [image.yaml](https://github.com/nmagnezi/assisted-service-crs/blob/cc900a4c39295c0b9204ff3bb25f5ed940f402a9/image.yaml)

**Result**

`kubectl -n assisted-installer describe cluster cluster-sample`
```yaml
Name:         cluster-sample
Namespace:    assisted-installer
Labels:       <none>
Annotations:  kubectl.kubernetes.io/last-applied-configuration:
                {"apiVersion":"adi.io.my.domain/v1alpha1","kind":"Cluster","metadata":{"annotations":{},"name":"cluster-sample","namespace":"assisted-inst...
API Version:  adi.io.my.domain/v1alpha1
Kind:         Cluster
Metadata:
  Creation Timestamp:  2021-02-01T15:11:21Z
  Generation:          1
  Managed Fields:
    API Version:  adi.io.my.domain/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .:
          f:kubectl.kubernetes.io/last-applied-configuration:
      f:spec:
        .:
        f:clusterNetworkCIDR:
        f:clusterNetworkHostPrefix:
        f:name:
        f:openshiftVersion:
        f:provisionRequirements:
          .:
          f:agentSelector:
            .:
            f:matchLabels:
              .:
              f:bla:
          f:controlPlaneAgents:
        f:pullSecretRef:
          .:
          f:name:
          f:namespace:
    Manager:      kubectl
    Operation:    Update
    Time:         2021-02-01T15:11:21Z
    API Version:  adi.io.my.domain/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:status:
        .:
        f:connectivityMajorityGroups:
        f:id:
        f:lastUpdateTime:
        f:progress:
        f:state:
        f:stateInfo:
        f:validationsInfo:
    Manager:         assisted-service
    Operation:       Update
    Time:            2021-02-01T15:11:27Z
  Resource Version:  12009
  UID:               046ac397-d34f-4198-86cc-13ec135a1523
Spec:
  Cluster Network CIDR:         10.128.0.0/14
  Cluster Network Host Prefix:  23
  Name:                         test-cluster
  Openshift Version:            4.7
  Provision Requirements:
    Agent Selector:
      Match Labels:
        Bla:               aaa
    Control Plane Agents:  3
  Pull Secret Ref:
    Name:       pull-secret
    Namespace:  assisted-installer
Status:
  Connectivity Majority Groups:  {}
  Id:                            02400a00-c94f-4bb4-84a4-c698a77e5258
  Last Update Time:              2021-02-01T15:21:15Z
  Progress:
  State:             pending-for-input
  State Info:        User input required
  Validations Info:  {"configuration":[{"id":"pull-secret-set","status":"success","message":"The pull secret is set."}],"hosts-data":[{"id":"all-hosts-are-ready-to-install","status":"success","message":"All hosts in the cluster are ready to install."},{"id":"sufficient-masters-count","status":"failure","message":"Clusters with less than 3 dedicated masters or a single worker are not supported. Please either add hosts, or disable the worker host"}],"network":[{"id":"machine-cidr-defined","status":"failure","message":"The Machine Network CIDR is undefined; the Machine Network CIDR can be defined by setting either the API or Ingress virtual IPs."},{"id":"machine-cidr-equals-to-calculated-cidr","status":"pending","message":"The Machine Network CIDR, API virtual IP, or Ingress virtual IP is undefined."},{"id":"api-vip-defined","status":"failure","message":"The API virtual IP is undefined and must be provided."},{"id":"api-vip-valid","status":"pending","message":"The API virtual IP is undefined."},{"id":"ingress-vip-defined","status":"failure","message":"The Ingress virtual IP is undefined and must be provided."},{"id":"ingress-vip-valid","status":"pending","message":"The Ingress virtual IP is undefined."},{"id":"dns-domain-defined","status":"failure","message":"The base domain is undefined and must be provided."},{"id":"cluster-cidr-defined","status":"success","message":"The Cluster Network CIDR is defined."},{"id":"service-cidr-defined","status":"failure","message":"The Service Network CIDR is undefined."},{"id":"no-cidrs-overlapping","status":"pending","message":"At least one of the CIDRs (Machine Network, Cluster Network, Service Network) is undefined."},{"id":"network-prefix-valid","status":"success","message":"The Cluster Network prefix is valid."},{"id":"ntp-server-configured","status":"success","message":"No ntp problems found"}]}
Events:              <none>
```

`kubectl -n assisted-installer describe images.adi.io.my.domain image-sample`
```yaml
Name:         image-sample
Namespace:    assisted-installer
Labels:       <none>
Annotations:  kubectl.kubernetes.io/last-applied-configuration:
                {"apiVersion":"adi.io.my.domain/v1alpha1","kind":"Image","metadata":{"annotations":{},"name":"image-sample","namespace":"assisted-installe...
API Version:  adi.io.my.domain/v1alpha1
Kind:         Image
Metadata:
  Creation Timestamp:  2021-02-01T15:12:17Z
  Generation:          1
  Managed Fields:
    API Version:  adi.io.my.domain/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .:
          f:kubectl.kubernetes.io/last-applied-configuration:
      f:spec:
        .:
        f:clusterRef:
          .:
          f:name:
          f:namespace:
        f:sshPublicKey:
    Manager:      kubectl
    Operation:    Update
    Time:         2021-02-01T15:12:17Z
    API Version:  adi.io.my.domain/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:status:
        .:
        f:expirationTime:
        f:sizeBytes:
        f:state:
    Manager:         assisted-service
    Operation:       Update
    Time:            2021-02-01T15:12:54Z
  Resource Version:  12176
  UID:               af72d112-8d5f-40c3-b756-e18d3f0dd23e
Spec:
  Cluster Ref:
    Name:          cluster-sample
    Namespace:     assisted-installer
  Ssh Public Key:  mykey
Status:
  Expiration Time:  2021-02-01T19:22:41Z
  Size Bytes:       946864128
  State:            Image has been created
Events:             <none>
```

File: internal/controller/controllers/image_controller.go [hunk 1]
```diff
@@ -67,6 +67,35 @@ func (r *ImageReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 }
 
 func (r *ImageReconciler) create(ctx context.Context, image *adiiov1alpha1.Image) (ctrl.Result, error) {
+	covertStaticIPConfig := func(crdStaticIPConfig []*adiiov1alpha1.StaticIPConfig) []*models.StaticIPConfig {
```
Comment (filanov) [inline internal/controller/controllers/image_controller.go:70]: this function is already long enough, should probably be outside as a helper function.

File: internal/controller/controllers/image_controller.go [hunk 2]
```diff
@@ -67,6 +67,35 @@ func (r *ImageReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
...
+		var newStaticIPsConfig []*models.StaticIPConfig
+		for i := 0; i < len(crdStaticIPConfig); i++ {
```
Comment (filanov) [inline internal/controller/controllers/image_controller.go:72]: More go like
```suggestion
		for i := range crdStaticIPConfig {
```

File: internal/controller/api/v1alpha1/image_types.go [hunk 3]
```diff
@@ -26,14 +26,20 @@ const (
...
 // ImageSpec defines the desired state of Image
 type ImageSpec struct {
 	// The name of the Cluster CR
 	ClusterRef   *ClusterReference `json:"clusterRef"`
 	SSHPublicKey string            `json:"sshPublicKey,omitempty"`
 	// The name of the secret containing the pull secret
-	IgnitionOverrides     string                   `json:"ignitionOverrides,omitempty"`
-	StaticIpConfiguration []*models.StaticIPConfig `json:"staticIpConfiguration,omitempty"`
+	IgnitionOverrides     string            `json:"ignitionOverrides,omitempty"`
+	StaticIpConfiguration []*StaticIPConfig `json:"staticIpConfiguration,omitempty"`
```
Comment (filanov) [inline internal/controller/api/v1alpha1/image_types.go:42]: maybe `staticIPConfiguration`?

File: internal/controller/controllers/image_controller.go (hunk 1 above)
Comment (nmagnezi) [reply internal/controller/controllers/image_controller.go]: Done

File: internal/controller/controllers/image_controller.go (hunk 2 above)
Comment (nmagnezi) [reply internal/controller/controllers/image_controller.go]: Done

File: internal/controller/api/v1alpha1/image_types.go (hunk 3 above)
Comment (nmagnezi) [reply internal/controller/api/v1alpha1/image_types.go]: Done

Comment (filanov) [reply internal/controller/api/v1alpha1/image_types.go]: I was talking about the json
```
StaticIpConfiguration []*StaticIPConfig `json:"staticIPConfiguration,omitempty"`
```

Comment (nmagnezi) [reply internal/controller/api/v1alpha1/image_types.go]: gotcha. done now 👍

**Review by filanov — APPROVED** [review]

**Pull request merged**